File: packages/repo-gc-shared/src/heuristics/implicit-control.ts
import { Finding, FindingKind, Severity, type FileData, type Thresholds } from '../types';

export function analyzeImplicitControl(
  data: FileData,
  thresholds: Thresholds,
  idCounter: { value: number },
): Finding | null {
  const decoratorCount = data.decoratorCount;
  const fnCount = Math.max(data.functionCount, 1);
  const ratio = decoratorCount / fnCount;

  // Framework-aware: NestJS/Angular/dataclass-heavy files get 1.5x threshold
  const effectiveLimit = isFrameworkHeavyFile(data)
    ? thresholds.decoratorDensityLimit * 1.5
    : thresholds.decoratorDensityLimit;

  if (ratio <= effectiveLimit) return null;

  let severity: Severity;
  if (ratio >= effectiveLimit * 3) {
    severity = Severity.High;
  } else if (ratio > effectiveLimit * 2) {
    severity = Severity.Medium;
  } else {
    severity = Severity.Low;
  }

  return {
    id: `ic-${String(idCounter.value++).padStart(3, '0')}`,
    kind: FindingKind.ImplicitControl,
    severity,
    confidence: 0.40,
    path: data.relativePath,
    summary: '',
    reasons: [],
    evidence: [
      `decorator_count: ${decoratorCount}`,
      `function_count: ${fnCount}`,
      `ratio: ${ratio.toFixed(3)}`,
      `limit: ${effectiveLimit.toFixed(3)}`,
    ],
    suggested_next_step: '',
    estimated_tokens: undefined,
  };
}

function isFrameworkHeavyFile(data: FileData): boolean {
  const frameworkPatterns = [
    /@nestjs\//, /@angular\//, /@Component/, /@NgModule/,
    /@Entity/, /@Injectable/, /@Controller/, /@Service/,
    /@dataclass/, /@pydantic/,
  ];
  for (const imp of data.imports) {
    for (const pat of frameworkPatterns) {
      if (pat.test(imp)) return true;
    }
  }
  return false;
}

File: packages/repo-gc-python/python/repo_gc_python/heuristics/implicit_control.py
"""Implicit control flow heuristic — detects high decorator/middleware density.

Decorators create hidden execution paths that LLMs may miss.
Framework-aware: dataclass/pydantic-heavy files get 1.5x threshold multiplier.
"""

from __future__ import annotations

from ..cli import Threshold
from ..discovery import PythonFile
from ..parsing import FileInfo
from ..types import Finding, FindingKind, Severity


def analyze(
    pfile: PythonFile,
    info: FileInfo,
    threshold: Threshold,
    counter: int,
) -> Finding | None:
    decorator_count = info.decorator_count
    fn_count = max(info.function_count, 1)
    ratio = decorator_count / fn_count

    # Framework-aware: dataclass/pydantic-heavy files get 1.5x threshold
    effective_limit = _effective_limit(threshold, info)

    if ratio <= effective_limit:
        return None

    if ratio >= effective_limit * 3:
        severity = Severity.High
    elif ratio > effective_limit * 2:
        severity = Severity.Medium
    else:
        severity = Severity.Low

    return Finding(
        id=f"ic-{counter:03d}",
        kind=FindingKind.ImplicitControl,
        severity=severity,
        confidence=0.40,
        path=str(pfile.relative_path),
        summary="",
        reasons=[],
        evidence=[
            f"decorator_count: {decorator_count}",
            f"function_count: {fn_count}",
            f"ratio: {ratio:.3f}",
            f"limit: {effective_limit:.3f}",
        ],
        suggested_next_step="",
        estimated_tokens=None,
    )


def _effective_limit(threshold: Threshold, info: FileInfo) -> float:
    """Apply framework-aware threshold multiplier (1.5x for known frameworks)."""
    base = threshold.decorator_density_limit
    framework_imports = [
        "dataclasses", "pydantic", "attr", "attrs",
        "sqlalchemy", "django", "flask", "fastapi",
        "typing_extensions",
    ]
    for imp in info.imported_modules:
        for fw in framework_imports:
            if fw in imp:
                return base * 1.5
    return base
